velox/type/Filter.cpp

```diff
@@ -458,6 +458,51 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  if (other->kind() == FilterKind::kMultiRange) {
+    bool bothNullAllowed = nullAllowed_ && other->testNull();
+    // TODO: verify nanAllowed with the other filter.
+    const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+    std::vector<std::unique_ptr<Filter>> merged;
+    for (auto const& filter : filters()) {
+      for (auto const& filterOther : multiRangeOther->filters()) {
+        auto innerMerged = filter->mergeWith(filterOther.get());
+        switch (innerMerged->kind()) {
+          case FilterKind::kAlwaysFalse:
+            break;
+          case FilterKind::kIsNotNull:
+          case FilterKind::kIsNull:
+            // the merged filter is valid when it matches with top-level
+            // nullAllowed.
+            if (innerMerged->testNull() && bothNullAllowed) {
+              merged.push_back(std::move(innerMerged));
+            }
+            break;
+          default:
+            merged.push_back(std::move(innerMerged));
+        }
+      }
+    }
+
+    if (merged.empty()) {
+      return nullOrFalse(bothNullAllowed);
+    } else if (merged.size() == 1) {
+      // if both sides allow null while the merged inner filter does not allow
```

Comment:
style: Use full sentences for comments, e.g. start with a capital letter.

Reply:
ack